### ISSUE TITLE
Upgrade to Chrome Dev 79 (pinned to 78)

### ISSUE THREAD

**Issue by foolip** [web-platform-tests/wpt#19367]
[Postmortem: Chrome Dev 78→79 broke PR and master runs](https://github.com/web-platform-tests/wpt/issues/19362) tracks breakage triggered by the Chrome Dev version bumping from 78 to 79.

https://github.com/web-platform-tests/wpt/pull/19360 pinned the version to 78, and this issue tracks getting it unpinned again.

@JohnChen0

**Comment by stephenmcgruer** [comment]
I've been trying to work on this in https://github.com/web-platform-tests/wpt/pull/19609. So far my main finding is that we seem to get stuck in our self.wait_script:

```
do_test for WebDriverRefTestExecutor
Got command: u'log'
RefTestImplementation: run_test. test.references: [(<wptrunner.wpttest.ReftestTest /infrastructure/assumptions/ahem-ref.html>, u'==')]
Got command: u'log'
screenshot: url=http://web-platform.test:8000/infrastructure/assumptions/ahem.html, timeout: 10
Got command: u'log'
_screenshot, executing self.wait_script...
<--- there should be more logs here, based on the CL that ran
Got command: 'test_ended'
Unexpected count in this thread 1
TEST-UNEXPECTED-TIMEOUT | /infrastructure/assumptions/ahem.html | expected PASS
```

I've not been able to figure out why yet.

Attempts to reproduce locally have failed, even using the docker image. For the record, this was my setup for docker WPT:

```
$ docker pull harjgam/web-platform-tests:0.33
$ docker run -it --privileged harjgam/web-platform-tests:0.33
# ./start.sh
# cd web-platform-tests
# git remote add stephenmcgruer https://github.com/stephenmcgruer/web-platform-tests
# git fetch stephenmcgruer
# git checkout stephenmcgruer/unpin-chrome-dev
// Note that run_tc.py REALLY doesn't work locally despite what the documentation says, so lets
// do it manually.
# sudo sh -c "./wpt make-hosts-file >> /etc/hosts"
# wget https://dl.google.com/linux/direct/google-chrome-unstable_current_amd64.deb
# sudo apt-get -qqy update
# sudo gdebi -qn google-chrome-unstable_current_amd64.deb
# sudo Xvfb $DISPLAY -screen 0 1280x1024x24
<ctrl-z>
# bg
# sudo fluxbox -display $DISPLAY
<ctrl-z>
# bg
# ./tools/ci/ci_wptrunner_infrastructure.sh
...
TEST-PASS | /infrastructure/assumptions/ahem.html | took 1133ms
```

Doing this seems as close to Taskcluster as I can get, but still passes :(

**Comment by stephenmcgruer** [comment]
[Log with chromedriver --verbose enabled](https://github.com/web-platform-tests/wpt/files/3714598/live_backing.log.txt). @JohnChen0 - from that log I'm pretty sure that Chromedriver sent the contents of [reftest-wait_webdriver.js](https://github.com/web-platform-tests/wpt/blob/221380dbd61b501ec6e580f7cfc0d43e8372064a/tools/wptrunner/wptrunner/executors/reftest-wait_webdriver.js) to Chrome, and just never received a reply?

**Comment by foolip** [comment]
If it seems like it's Chrome that's not well behaved, then bisecting Chrome while holding ChromeDriver constant should be pretty doable. If only it would repro locally, that is...

In Taskcluster, you could try to pin the ChromeDriver version to the old or new version, to figure out if it's Chrome or ChromeDriver that's gone bad. After that... I guess one could bisect in a very slow manner on Taskcluster if there are no other clues to be found.

**Comment by stephenmcgruer** [comment]
That's pretty much my next idea :). I'm going to start by getting logs out of Chrome, should be fairly easy to pump them to a file and then dump the file after the run. If that doesn't bear fruit, a bit of bisecting will be so much fun... :D

**Comment by stephenmcgruer** [comment]
Woah, the Chrome logs on the failing taskcluster are interesting...

```
[1433:1455:1011/125943.821785:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
[1471:1471:1011/125944.019461:ERROR:sandbox_linux.cc(369)] InitializeSandbox() called with multiple threads in process gpu-process.
[1433:1556:1011/125944.088840:ERROR:bus.cc(393)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
[1433:1556:1011/125944.132266:ERROR:bus.cc(393)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
[1433:1556:1011/125944.132534:ERROR:bus.cc(393)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
[1433:1580:1011/125944.299323:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
[1433:1580:1011/125944.299580:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
[1433:1580:1011/125944.299796:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
[1433:1580:1011/125944.299959:WARNING:property.cc(145)] DaemonVersion: GetAndBlock: failed.
[1433:1580:1011/125944.300006:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
[1433:1580:1011/125944.300194:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
[1471:1471:1011/125944.444046:WARNING:x11_util.cc(1415)] X error received: serial 325, error_code 128 (BadShmSeg (invalid shared segment parameter)), request_code 130, minor_code 3 (X_ShmPutImage)
[1433:1433:1011/125944.950480:INFO:CONSOLE(50)] "observer.observe(document.documentElement, {attributes: true});", source:  (50)
[1433:1433:1011/125944.959206:INFO:CONSOLE(53)] "document.readyState: complete", source:  (53)
[1433:1433:1011/125944.961085:INFO:CONSOLE(58)] "call root_wait()", source:  (58)
[1433:1433:1011/125944.963253:INFO:CONSOLE(4)] "root.classList.contains("reftest-wait"): false", source:  (4)
[1433:1433:1011/125944.965253:INFO:CONSOLE(8)] "disconnected observer", source:  (8)
[1433:1433:1011/125944.967229:INFO:CONSOLE(10)] "Document.prototype.hasOwnProperty("fonts"): true", source:  (10)
[1433:1433:1011/125944.969218:INFO:CONSOLE(12)] "waiting on document.fonts.ready", source:  (12)
[1433:1433:1011/125944.972163:INFO:CONSOLE(14)] "document.fonts.ready fired, calling ready_for_screenshot", source:  (14)
[1433:1433:1011/125944.973249:INFO:CONSOLE(28)] "in ready_for_screenshot()", source:  (28)
[1433:1450:1011/130000.020259:WARNING:gpu_process_host.cc(1202)] The GPU process has crashed 1 time(s)
[1433:1433:1011/130000.037483:ERROR:zygote_communication_linux.cc(276)] Failed to send GetTerminationStatus message to zygote
[1433:1433:1011/130000.037535:WARNING:zygote_communication_linux.cc(288)] Socket closed prematurely.
[1433:1433:1011/130000.039811:ERROR:zygote_communication_linux.cc(276)] Failed to send GetTerminationStatus message to zygote
[1433:1433:1011/130000.039832:WARNING:zygote_communication_linux.cc(288)] Socket closed prematurely.
```

A few notes:

* The logs encode date/time; the format is `process_id:thread_id:month/dayhhmmss.fractions`
* Therefore, we can see that from the first log to the last, the self.wait_for script takes only a few fractions of a second. Cross-referencing with Chromedriver logs (which, with `--verbose`, also log JS console logging as it turns out), we can see that the script starts nearly immediately after the python code calls it.
* The last call is before **waiting for rAF**. `ready_for_screenshot()` immediately calls `requestAnimationFrame`. I have no idea how rAF could not be happening...

**Comment by stephenmcgruer** [comment]
This gets stranger and stranger. I hacked `ready_for_screenshot()` to just call `callback()` immediately and chromedriver fails to take a screenshot o.o.

```
PID 1422 | [1570800823.550][DEBUG]: DevTools WebSocket Event: Runtime.consoleAPICalled 087D7A4D0328CE6EE21138F09B366D4F {
PID 1422 |    "args": [ {
PID 1422 |       "type": "string",
PID 1422 |       "value": "hack; instead of rAF, immediately callback()"
...
_screenshot, executing webdriver.screenshot()
...
PID 1422 | [1570800823.563][INFO]: [a59b74f88b5010c31bac1726286a0e18] RESPONSE ExecuteAsyncScript null
PID 1422 | [1570800823.564][INFO]: [a59b74f88b5010c31bac1726286a0e18] COMMAND Screenshot {
...
PID 1422 | [1570800823.572][DEBUG]: DevTools WebSocket Command: Page.captureScreenshot (id=42) 087D7A4D0328CE6EE21138F09B366D4F {
PID 1422 |    "clip": {
...
PID 1422 | [1570800833.572][SEVERE]: Timed out receiving message from renderer: 9.999
PID 1422 | [1570800833.577][WARNING]: screenshot failed, retrying
PID 1422 | [1570800833.577][DEBUG]: DevTools WebSocket Command: Page.captureScreenshot (id=43) 087D7A4D0328CE6EE21138F09B366D4F {
PID 1422 |    "clip": {
...
Got command: 'test_ended'
Unexpected count in this thread 1
TEST-UNEXPECTED-TIMEOUT | /infrastructure/assumptions/ahem.html | Waiting on browser:
```

So we have some weird thing where rAF doesn't appear to be firing, and some weird thing where Page.captureScreenshot just isn't working.

I think it's time to bisect :(. This may be painful.

**Comment by stephenmcgruer** [comment]
Bisected chrome at a pinned Chromedriver version;

```
Revision 694908 is [(g)ood/(b)ad/(r)etry/(u)nknown/(s)tdout/(q)uit]: b
You are probably looking for a change made after 694904 (known good), but no later than 694908 (first known bad).
CHANGELOG URL:
  https://chromium.googlesource.com/chromium/src/+log/af8e5c0e26dcb4ba200ea0664065b1cc45225bd1..b4f3df4e77ba1f79717d44b5068cb36f7537332f
```

**Comment by stephenmcgruer** [comment]
From   https://chromium.googlesource.com/chromium/src/+log/af8e5c0e26dcb4ba200ea0664065b1cc45225bd1..b4f3df4e77ba1f79717d44b5068cb36f7537332f, I would suspect https://chromium.googlesource.com/chromium/src/+/b4f3df4e77ba1f79717d44b5068cb36f7537332f

Note that on a number of runs I saw: `[491:502:1011/191621.730263:WARNING:gpu_process_host.cc(1218)] The GPU process has crashed 1 time(s)`

I'm going to build a ToT Chromium with that reverted, put it somewhere taskcluster can find it, and see if I can prove that CL is responsible. If so, I'll open a crbug and we'll move onto the next step in debugging this :D

**Comment by foolip** [comment]
Excellent and exciting debugging there, hope it bears fruit!

**Comment by stephenmcgruer** [comment]
A little note to make sure I dont forget:

> Note that on a number of runs I saw: [491:502:1011/191621.730263:WARNING:gpu_process_host.cc(1218)] The GPU process has crashed 1 time(s)

I see this on successful runs too, as it turns out. Likely a red-herring.

**Comment by stephenmcgruer** [comment]
Filed https://crbug.com/1014629

**Comment by stephenmcgruer** [comment]
Update: a fix was developed and landed in Chromium, it will be in the `79.0.3945.10` release. It is expected to come on the dev channel sometime next week.

**Comment by foolip** [comment]
Excellent, thanks @stephenmcgruer!
